src/librustc/infer/error_reporting/mod.rs

```diff
@@ -1097,7 +1097,9 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
             if let Some((sp, has_lifetimes)) = type_param_span {
                 let tail = if has_lifetimes { " + " } else { "" };
                 let suggestion = format!("{}: {}{}", bound_kind, sub, tail);
-                err.span_suggestion_short(sp, consider, suggestion);
+                err.span_suggestion_short_with_applicability(
+                    sp, consider, suggestion, Applicability::MaybeIncorrect
+                );
```

Comment:
In most cases, the explanation is just "I don't feel Overwhelmingly Confident that this suggestion will always be what the user wants, even if I don't have an explicit counterexample of reasonable code for which this suggestion is incorrect", because I was imagining `MachineApplicability` was a very high standard. ([Like you said](https://github.com/rust-lang/rust/pull/50476#issuecomment-388239644), we don't want rustfix to guess when it doesn't _know_.)

If the standard is slightly lower than that (such that `MaybeIncorrect` requires a commentable justification stronger than "Well, I'm not _sure_ that it's always correct"), then I'd want to change a lot of these to `MachineApplicable`. Thinking on it, that's probably the best long-term decision, for the same reason assertions and static typing are good ideas: better to make slightly too strong assumptions (and then fix the bugs as they're reported), rather than too weak assumptions (and have no way of knowing whether we could successfully auto-fix more things).

Reply:
I feel that having the _convention_ of having a comment outlining the reasons the suggestion might be incorrect could be a good start. We can _in the future_ encode this using types.